File: 0020_Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool: 
        stack = ['N']
        map = {'N': 'N', '{': '}', '(': ')', '[': ']'}
        for char in s:
            if char in map: # open paren
                stack.append(char)
            elif map[stack.pop()] != char:
                return False

        return stack == ['N']

File: 0020_Valid_parentheses.cpp
#include <string>
#include <stack>

class Solution {
public:
    bool isValid(std::string s) {
        std::stack<char> st;

        for (char c: s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            }
            if (c == '}') {
                if (st.empty() || st.top() != '{')
                    return false;
                st.pop();
            }
            if (c == ')') {
                if (st.empty() || st.top() != '(')
                    return false;
                st.pop();
            }
            if (c == ']') {
                if (st.empty() || st.top() != '[')
                    return false;
                st.pop();
            }
        }
        if (st.empty())
            return true;
        return false;
    }
};
